### PR TITLE
ggml : do not use _GNU_SOURCE gratuitously

### PR DESCRIPTION
**Pull request by ggerganov**
See https://github.com/ggerganov/whisper.cpp/pull/1027

### CODE REVIEW

**Comment by ggerganov** [comment]
@przemoc

I'm trying to apply the _GNU_SOURCE patch from `whisper.cpp` but hitting the following error on MacOS:

```gcc
In file included from llama.cpp:9:
./llama-util.h:189:63: error: use of undeclared identifier 'MADV_WILLNEED'
            if (madvise(addr, std::min(file->size, prefetch), MADV_WILLNEED)) {
                                                              ^
./llama-util.h:316:38: error: use of undeclared identifier 'RLIMIT_MEMLOCK'
            if (suggest && getrlimit(RLIMIT_MEMLOCK, &lock_limit))
                                     ^
2 errors generated.
make: *** [llama.o] Error 1
Error: Process completed with exit code 2.
```

Any ideas how to resolve?

**Comment by przemoc** [comment]
Sorry, I will be momentarily afk for rest of the day, so cannot help much.

I would need to look into llama.cpp, but `madvise()` typically can be replaced with more portable [posix_madvise()](https://pubs.opengroup.org/onlinepubs/009695399/functions/posix_madvise.html) and `MADV_WILLNEED` with analogous `POSIX_MADV_WILLNEED` (unless you sometimes use in the code also `advice` that is not defined by POSIX.1).

`getrlimit()` is slightly different case, because you use resource `RLIMIT_MEMLOCK` that is not defined by POSIX.1 (even if you would look at the latest POSIX.1-2017), [it comes from BSD](https://man7.org/linux/man-pages/man2/getrlimit.2.html#STANDARDS). You may check if having `_BSD_SOURCE` additionally defined (`-D_BSD_SOURCE` after already defined `-D_XOPEN_SOURCE=600`) helps for macOS.

File: ggml.c [hunk 1]
```diff
@@ -13387,8 +13386,7 @@ static void ggml_compute_forward_conv_2d_sk_p0_f16_f32(
     const int nk1 = ne01;
 
     // size of the convolution row - the kernel size unrolled across all channels
-    // round-up so it is more suitable for SIMD
-    const int ew0 = ggml_up32(nk0*nk1*ne02);
+    const int ew0 = nk0*nk1*ne02;
```
Comment (przemoc) [inline ggml.c:13389]: This change is unrelated to `_GNU_SOURCE` removal, so shouldn't be part of this PR.

**Comment by przemoc** [comment]
@ggerganov, see my attempt for `llama.cpp` in #2035. Please run CI checks there, because for macOS I'm kind of working in the dark. ;)

**Comment by przemoc** [comment]
As we're dealing with this topic in my PR #2035 for `llama.cpp`, I think you may close your own PR #1992 (that was based on my initial attempt at `whisper.cpp`) to declutter PR space. :)
